pkg/util/containers/collectors/detector.go

```diff
@@ -0,0 +1,145 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2018 Datadog, Inc.
+
+package collectors
+
+import (
+	"errors"
+	"strings"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/datadog-agent/pkg/util/retry"
+)
+
+// ErrNothingYet is returned when no collector is currently detected.
+// This might change in the future if new collectors are valid.
+var ErrNothingYet = &retry.Error{
+	LogicError:    errors.New("No collector detected"),
+	RessourceName: "detector",
+	RetryStatus:   retry.FailWillRetry,
+}
+
+// ErrPermaFail is returned when all candidates have permanently failed.
+var ErrPermaFail = &retry.Error{
+	LogicError:    errors.New("No collector available"),
+	RessourceName: "detector",
+	RetryStatus:   retry.PermaFail,
+}
+
+// Detector holds the logic to initialise collectors, with retries,
+// and selecting the most appropriate among available collectors
+type Detector struct {
+	candidates         map[string]Collector
+	detected           map[string]Collector
+	preferredCollector Collector
+	preferredName      string
+}
+
+// NewDetector returns a Detector ready to use. If configuredName
+// is empty, autodetection is enabled. If not, only the one name
+// will be tried.
+func NewDetector(configuredName string) *Detector {
+	d := &Detector{
+		candidates: make(map[string]Collector),
+		detected:   make(map[string]Collector),
+	}
+	// Load candidates from catalog
+	for n, f := range defaultCatalog {
+		if configuredName != "" && n != configuredName {
+			// If a collector name is configured, skip the others
+			continue
+		}
+		d.candidates[n] = f()
+	}
+	return d
+}
+
+// GetPreferred detects, ranks and returns the best collector for now.
+// Result might change if new collectors are valid after start, then
+// constant when all collectors are either ok or PermaFail.
+// Users should keep calling this method instead of caching the first result.
+func (d *Detector) GetPreferred() (Collector, string, error) {
+	if d.candidates == nil {
+		// Detection rounds finished, exit fast
+		if d.preferredCollector == nil {
+			return nil, "", ErrPermaFail
+		}
+		return d.preferredCollector, d.preferredName, nil
+	}
+
+	d.detectCandidates()
+	// Stop detection when all candidates are tested
+	if len(d.candidates) == 0 {
+		d.candidates = nil
+		d.detected = nil
+	}
+
+	if d.preferredCollector == nil {
+		if d.candidates == nil {
+			return nil, "", ErrPermaFail
```

Comment:
the combination of `d.preferredCollector == nil` and `d.candidates == nil` should already have been returned before (line 67)

Reply:
Not when the last candidate is removed (either ok or permafail). Not having that here would mean we return `ErrNothingYet` that time instead of `ErrPermaFail`